File: wordBreak_dp.py
"""
TC=O(N^3) O(N)-->iteration for i and O(N) for j--i and O(N ) for substr
SC=O(N+1) -->dp table
"""


class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        if(s is None or len(s)==0):
            return False
        dp=[False]*(len(s)+1)
        dp[0]=True
        
        for i in range(len(dp)):
            for j in range(0,i):
                if(dp[j] and s[j:i] in wordDict):
                    dp[i]=True
                    break
        return dp[-1]
                    

File: word-break-dp.cpp
// Time Complexity : O(n^3)
// Space Complexity : O(nL)
// Did this code successfully run on Leetcode : NO (Time Limit Exceeded)

#include <iostream>
#include <vector>
#include <unordered_set>

using namespace std;

//dp
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        unordered_set<string> set;
        for(auto it: wordDict){
            set.insert(it);
        }
        vector<bool> dp(s.size() + 1);
        dp[0] =  true;
        for(int i = 1; i < dp.size(); i++){
            for(int j = 0; j < i; j++){
                if(dp[j] && set.find(s.substr(j, i - j)) != set.end()){
                    dp[i] = true;
                }
            }
        }
        return dp[dp.size() - 1];
    }
};